### ISSUE TITLE
ZenBot not buying/selling

### ISSUE THREAD

**Issue by tenpaiyomi** [DeviaVir/zenbot#868]
I did the initial run for a bit in paper mode to watch and get an idea of how it works.  Satisfied with that, I removed the `--paper` command from the dockerfile and docker-compose.yml files and started the container back up.  I see it say it's running in live mode with "STARTING LIVE TRADING", I see it saying "buy" and "sell" at various points, but....it does nothing.  If I explicitly issue a market buy or market sell, it will output market buy/sell, but never actually does a buy or sell.

I've run in debug mode with `--debug` and that doesn't actually seem to have any effect either, nor show any real debugging output.

This is with gdax BTC_USD.  Anything I might be missing?

**Comment by brucetus** [comment]
Try setting buy/sell pct to 99 and make sure your balance is above the amount necessary to open an order higher than the minimum order size on gdax.

**Comment by tenpaiyomi** [comment]
@brucetus They are already at 99%, and my account balance is $150, which I've been able to use to buy and sell at full amount without issue.  In fact, doing it manually I've made $10 profit which, while a small sum, still shows that it should be doable via zenbot.

**Comment by mhertis** [comment]
Same problem here on bitfinex.. manually executing orders on bitfinex web app works without a problem.
![image](https://user-images.githubusercontent.com/6152898/34312570-6adbbc4e-e765-11e7-88f5-f799e8b36a93.png)

**Comment by Deraen** [comment]
I think the problem (or a problem, anyway) is that e.g. [gdax/products.json](../blob/master/extensions/exchanges/gdax/products.json) defines minimum order size for GDAX BTC/EUR as 0.01 BTC though the real mimimum is 0.0001 BTC: https://support.gdax.com/customer/portal/articles/2725970-trading-rules (under order minimums). Files for other exchanges might also use wrong values.

I modified the file and got simulation working with a small capital value.

**Comment by DeviaVir** [comment]
This issue was moved to highvelocityspace/zenbot#80

**Comment by benjamincburns** [comment]
I did some sniffing 'round the code on this issue this evening by adding a bunch of verbose debug messages, and it seems there are a number of potential issues contributing to flaky order execution.

The ones I've found so far:

* It appears that there are a number of paths through `executeSignal` which can cause `s.acted_on_trend` to be set to `true` even though no order was placed, such as when some spurious network error occurs. This causes the signal to not be set on the next `onPeriod`, which means you'll miss the trend entirely if you don't make it through to `placeOrder` successfully on the first go around.
* I occasionally see order rejections due to insufficient balance, implying a rounding error in the calculation of `order.size` or `order.price` in `executeSignal` within `lib/engine.js`. Edit: The earliest indication of this that I've found so far is #487, but it seems this is a long standing issue based on the advice around not setting `markup_pct` to 100.
* I occasionally see order rejections due to post only errors, even during times when the volume/price movement is very slow, again implying rounding errors in `executeSignal`

I've taken a crack at fixing some of these problems, but my changes are hacks upon hacks at this point. This code should be using a proper fixed precision number library for price/size calculations, and `executeSignal` really needs to be teased apart. It could be a _lot_ cleaner if it were refactored to use promises and `async`/`await`.
